ehrql/assurance.py

```diff
@@ -69,11 +80,31 @@ def validate(variable_definitions, test_data):
     }
 
     # Validate results of query
-    return {
+    test_validation_errors = {
         patient_id: validation_result
         for patient_id, patient in test_data.items()
         if (validation_result := validate_patient(patient_id, patient, query_results))
     }
+    return {
+        "constraint_validation_errors": constraint_validation_errors,
+        "test_validation_errors": test_validation_errors,
+    }
+
+
+def validate_constraints(patient_id, records, table):
+    errors = []
+    for record in records:
+        for column, value in record.items():
+            constraints = table.schema.schema[column].constraints
+            for constraint in constraints:
+                is_valid = constraint.validate(value)
+                if not is_valid:
+                    error_msg = f"Validate test data: Found errors for patient {patient_id}. Column '{column}' did not meet the '{constraint}' constraint, got '{value}'."
+                    errors.append(error_msg)
+    if len(errors) == 0:
+        return errors
+    else:
+        raise DataValidationError(error_msg)
```

Comment:
As you said in Slack, raising an exception here means that users would have to fix constraint validation errors one by one.

So instead of raising an exception, why not just return `errors`?  You're already (on lines 66-8) collecting all the constraint validation errors.

The question then becomes what to do with the collected errors.  I should think you could separate out the `present` function into two parts: one that presents any constraint errors, and one that presents any test errors.

Reply:
thx! instead of returning an `DataValidationError`, we have decided to return both (1) all data validation errors and (2) all result validation errors in a message.